### ISSUE TITLE
Problem with key-up events on Windows

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#2119]
Reported on the mailing list.

> On my windows machine (Windows 7), when I hit 1 key which was in a separate numeric pad, 
>Gesture log window showed  :
>
>  T{ key-down { sym “1” } }
>  User input: 1 
>
>per one hit. The other hand, when I hit normal 1 key, it showed :
>
>  T{ key-down { sym “1” } }
>  User input: 1 
>  T{ key-up { sym “1” } }
>
>per one hit.

Information about their configuration:

> Windows 7 Home Premium
> Service Pack 1 32bit 
> 
> Japanese 
> 
> Japanese PS/2 keyboard (109) 
> 
> 32bit Factor  factor-windows-x86-32-0.98

And report of bug on a different system:

> I confirm the same behavior on Windows 10, latest development build of 64-bit Factor.
> It doesn't depend on the numpad or on physical keys.
> 
> For example, when I use the normal English/US layout, the keys for semicolon and comma (";" and ",") generate only the "key-down" events. Pressing "q" and "w" generates both "key-down" and "key-up".
> If I switch to the Programmer Dvorak layout, the physical keys that used to be "q" and "w" in the US layout now will produce the ";" and "," correspondingly, and they will again produce only the "key-down" events in the gesture-logger. Pressing the keys that formerly produced only "key-downs" as ";" and "," will now produce both "key-down" and "key-up", being mapped to "s" and "w".
> 
> It seems that the physical keyboard or the layout is not to blame. For some reason Factor eats some of the key-up events for punctuation characters.
> 
> There are many such characters that I can see treated differently by Factor: #&$^-=\`:;~/'"?@_[]{}(*+
These Factor's treatment of these keys is not dependent on the physical key, but on the logically mapped character it produces under the currently selected layout.
> 
> Numpad keys may be different in that even when Programmer Dvorak produces "a"-"f" or "x", they still don't give a ""key-up" event.

**Comment by AlexIljin** [comment]
Related issue: #1565.

**Comment by kusumotonorio** [comment]
I wrote the code for both this issue and #1565.
https://github.com/factor/factor/issues/1565#issuecomment-479512182 

It ignores caps-lock at key-down and key-up events. And number keys on the numeric pad,  ",", ".", etc provide key-up events.

https://github.com/kusumotonorio/factor/blob/windows-kn-plan2/basis/ui/backend/windows/windows.factor

https://github.com/factor/factor/compare/master...kusumotonorio:windows-kn-plan2

**Comment by kusumotonorio** [comment]
I wrote a new code.

The problem with the Windows backend is that when key down or key up events occur, 
we can not know for sure how it is actually printed by the key.

I did various trials and errors, but I was not able to realize the same functions as other platforms. So I changed my mind and I decided to ask the user to use what kind of keyboard he was using.

The original code can not be used in combination with the control key or alt key except for the alphabet and number keys, and the key up event does not occur with the numeric keys of the numeric keypad or symbol keys such as & and *.
The new code by default only generates the same key-down / key-up events as the original, and behaves the same except ignoring the state of the keycaps.

In the new code, if you are using a US keyboard, when you write

USE: ui.backend.windows.keyboard.101-104

in .factor-rc, you can solve the problem of the above original code.
Also, if you are using a  Japanese keyboard, when

USE: ui.backend.windows.keyboard.106-109

is written, windows backend works in accordance with the keyboard.

If I can confirm that this code works stable, I will make a pull request. 
I'm looking forward to opinion from you.


https://github.com/factor/factor/compare/master...kusumotonorio:windows-kn-plan2

**Comment by mrjbq7** [comment]
I wonder if there is a Windows API to get the type of keyboard. Since usually the OS asks as well.

**Comment by kusumotonorio** [comment]
Thank you for your idea, John. It sounds nice!

How can I use GetKeyboardType in my code?


--
KUSUMOTO Norio



> 2019/04/22 23:03、John Benediktsson <notifications@github.com>のメール:
> 
> I wonder if there is a Windows API to get the type of keyboard. Since usually the OS asks as well. 
> 
>

**Comment by AlexIljin** [comment]
Hello!  I've created a PR to add this function to windows.user32: https://github.com/factor/factor/pull/2133  When it's merged, you can do```USE: windows.user320 GetKeyboardType .1 GetKeyboardType .2 GetKeyboardType .``` 22.04.2019, 17:16, "kusumotonorio" <notifications@github.com>:Thank you for your idea, John. It sounds nice!How can I use GetKeyboardType in my code?--KUSUMOTO Norio> 2019/04/22 23:03、John Benediktsson <notifications@github.com>のメール:>> I wonder if there is a Windows API to get the type of keyboard. Since usually the OS asks as well.>>> > On Apr 22, 2019, at 5:29 AM, kusumotonorio <notifications@github.com> wrote:> >> > I wrote a new code.> >> > The problem with the Windows backend is that when key down or key up events occur,> > we can not know for sure how it is actually printed by the key.> >> > I did various trials and errors, but I was not able to realize the same functions as other platforms. So I changed my mind and I decided to ask the user to use what kind of keyboard he was using.> >> > The original code can not be used in combination with the control key or alt key except for the alphabet and number keys, and the key up event does not occur with the numeric keys of the numeric keypad or symbol keys such as & and *.> > The new code by default only generates the same key-down / key-up events as the original, and behaves the same except ignoring the state of the keycaps.> >> > In the new code, if you are using a US keyboard, when you write> >> > USE: ui.backend.windows.keyboard.101-104> >> > in .factor-rc, you can solve the problem of the above original code.> > Also, if you are using a Japanese keyboard, when> >> > USE: ui.backend.windows.keyboard.106-109> >> > is written, windows backend works in accordance with the keyboard.> >> > If I can confirm that this code works stable, I will make a pull request.> > I'm looking forward to opinion from you.> >> > master...kusumotonorio:windows-kn-plan2> >> > —> > You are receiving this because you authored the thread.> > Reply to this email directly, view it on GitHub, or mute the thread.> —> You are receiving this because you commented.> Reply to this email directly, view it on GitHub, or mute the thread.> —You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.  ---=====---Александр

**Comment by AlexIljin** [comment]
Interestingly, my keyboard reports the following:
```
0 2 [a,b] [ GetKeyboardType . ] each
7
0
12
```

I could have sworn I'm not using a Japanese keyboard, but here we are.

**Comment by kusumotonorio** [comment]
Thank you, Alexander. 
Hmm, That's strange.

**Comment by kusumotonorio** [comment]
I tried.

```    0 GetKeyboardType {
        { 0x01 [ 101-US-keydown-codes ] } ! IBM PC/XT 83
        { 0x02 [ 101-US-keydown-codes ] } ! Olivetti ICO 102
        { 0x03 [ 101-US-keydown-codes ] } ! IBM PC/AT 84
        { 0x04 [ 101-US-keydown-codes ] } ! IBM enhanced 101 or 102
        { 0x05 [ 101-US-keydown-codes ] } ! Nokia 1050
        { 0x06 [ 101-US-keydown-codes ] } ! Nokia 9140
        { 0x07 [                          ! Japanese
              1 GetKeyboardType {
                  { 0x00 [ 101-US-keydown-codes ] } ! PC/AT 101 Enhanced 
                  { 0x01 [ 101-keydown-codes ] }    ! AX compatible
                  { 0x02 [ 106-keydown-codes ] }    ! IBM PC/AT 106
                  { 0x03 [ 101-US-keydown-codes ] } ! IBM 5576-002/003
                  { 0x04 [ 101-US-keydown-codes ] } ! IBM 5576-001
                  { 0x11 [ 101-keydown-codes ] }    ! AX
                  [ drop 101-US-keydown-codes ]     ! others
              } case
          ] }
        [ drop 101-US-keydown-codes ]     ! others
    } case ;
```
This idea works well for me, a typical Japanese, but I wonder if it is for everyone. I don't know if classification by return value of this function works well.
For example, the UK, France and Italy keyboards are different. I don't know it is really possible to divide by this function.

http://kbdlayout.info/KBDUK/
http://kbdlayout.info/0000040c/
http://kbdlayout.info/00000410/

What really need is to know what kind of printing will be done / was done by the key when a key down / key up event occurs.
the "toAscii" function may answer that. However, I don't know how to receive the answer from that function. The function puts the answer in a buffer. Factor needs to prepare the buffer, pass the address, and retrieve the result.

https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-toascii

I lack the ability to use the C language interface.

**Comment by AlexIljin** [comment]
How did you plan to use this function? For example, where would you get the contents for the `*lpKeyState` parameter?

**Comment by kusumotonorio** [comment]
I will probably need to get it using GetKeyboardState. If I write using C, it will be like this:

```
BYTE keyboardState[256];
GetKeyboardState(keyboardState);
```

**Comment by kusumotonorio** [comment]
Apparently, the new code seems to be working well.
```
: key-sym ( wParam -- string/f )
    wm-keydown-codes at ; inline

:: handle-wm-keydown ( hWnd uMsg wParam lParam -- )
    wParam exclude-key-wm-keydown? [
        wParam key-sym [
            t hWnd send-key-down
        ] [
            256 <byte-array> :> key-state
            2 <byte-array> :> chars
            lParam 16 shift 0xff bitand :> scan-code
            key-state GetKeyboardState 0 = not [
                0 VK_CONTROL key-state set-nth
                0 VK_CAPITAL key-state set-nth
                ! 0 VK_LCONTROL lpKeyState set-nth
                ! 0 VK_RCONTROL lpKeyState set-nth
                wParam scan-code key-state chars 0 ToAscii 1 = [
                    chars first 1string f hWnd send-key-down
                ] when
            ] when
        ] if*
    ] unless ;
```
I will continue the test.

**Comment by mrjbq7** [comment]
Awesome! Thanks so much for taking time to figure out this issue. Maybe when you are ready, let us know latest version for others to test. 

Thanks!

**Comment by kusumotonorio** [comment]
It is going to be good because of the discussion with you and Alexander. 
Thank you.

**Comment by kusumotonorio** [comment]
Please check [665e2c5](https://github.com/factor/factor/commit/665e2c582ed81663bb383ae8ad452e8b65d44e12) test.

This version supports the  dead code combination input and AltGr combination input required by European languages. I hope it's code that can be merged.


https://github.com/factor/factor/compare/master...kusumotonorio:windows-kn-plan2

**Comment by erg** [comment]
This issue is resolved in 08aa27a112aebede856ae6c94e2d09b65093b0c4
